flocker/control/test/test_httpapi.py

```diff
@@ -440,6 +442,289 @@ def created(ignored):
+
+        This is an alternative to the behaviour above.
+        The dataset creation API currently behaves this way.
+        Perhaps it shouldn't.
+        And instead allow datasets to be created with an as yet unknown node.
+        """
+        expected_manifestation = _manifestation()
+        node_a = Node(
+            hostname=self.NODE_A,
+            applications=frozenset(),
+            other_manifestations=frozenset([expected_manifestation])
+        )
+        deployment = Deployment(nodes=frozenset([node_a]))
+        saving = self.persistence_service.save(deployment)
+
+        def saved(ignored):
+            creating = self.assertResult(
+                b"POST",
+                b"/configuration/datasets/%s" % (
+                    expected_manifestation.dataset.dataset_id.encode('ascii')
+                ),
+                {u"primary": self.NODE_B},
+                BAD_REQUEST, {
+                    u"description":
+                    u"The provided primary node is not part of the cluster."
+                }
+            )
+            return creating
+        saving.addCallback(saved)
+        return saving
+    test_unknown_primary_node.todo = (
+        "See FLOC-1278.  Make this pass by inspecting cluster state "
+        "instead of desired configuration to determine whether a node is "
+        "valid or not."
+    )
+
+    def test_change_primary_to_configured_node(self):
+        """
+        If a different primary IP address is supplied and it identifies a node
+        which is already part of the cluster configuration, the modification
+        request succeeds and the dataset's primary becomes the given address.
+        """
+        expected_manifestation = _manifestation()
+        node_a = Node(
+            hostname=self.NODE_A,
+            applications=frozenset(),
+            other_manifestations=frozenset([expected_manifestation])
+        )
+        node_b = Node(hostname=self.NODE_B)
+        deployment = Deployment(nodes=frozenset([node_a, node_b]))
+        return self._test_change_primary(
+            expected_manifestation.dataset, deployment,
+            self.NODE_A, self.NODE_B
+        )
+
+    def test_primary_unchanged(self):
+        """
+        If the current primary IP address is supplied, the modification request
+        succeeds.
+        """
+        expected_manifestation = _manifestation()
+        node_a = Node(
+            hostname=self.NODE_A,
+            applications=frozenset(),
+            other_manifestations=frozenset([expected_manifestation])
+        )
+        node_b = Node(hostname=self.NODE_B)
+        deployment = Deployment(nodes=frozenset([node_a, node_b]))
+        return self._test_change_primary(
+            expected_manifestation.dataset, deployment,
+            self.NODE_A, self.NODE_A
+        )
+
+    def test_only_replicas(self):
+        """
+        If there are only replica manifestations of the requested dataset, 500
```

Comment:
As mentioned elsewhere, not having a primary will be a valid case in the future.